packages/app/src/cli/commands/app/dev.ts

```diff
@@ -139,46 +141,22 @@ If you're using the Ruby app template, then you need to complete the following s
       await showApiKeyDeprecationWarning()
     }
 
-    let tunnel: TunnelMode = {mode: 'auto'}
-    let tunnelType = 'cloudflare'
-
-    if (flags['use-localhost']) {
-      tunnelType = 'use-localhost'
-      tunnel = {
-        mode: 'no-tunnel-use-localhost',
-        provideCertificate: async (appDirectory) => {
-          renderInfo({
-            headline: 'Localhost-based development is in developer preview.',
-            body: [
-              '`--use-localhost` is not compatible with Shopify features which directly invoke your app',
-              '(such as Webhooks, App proxy, and Flow actions), or those which require testing your app from another',
-              'device (such as POS). Please report any issues and provide feedback on the dev community:',
-            ],
-            link: {
-              label: 'Create a feedback post',
-              url: 'https://community.shopify.dev/new-topic?category=shopify-cli-libraries&tags=app-dev-on-localhost',
-            },
-          })
-
-          return generateCertificate({
-            appDirectory,
-            onRequiresConfirmation: generateCertificatePrompt,
-          })
-        },
-      }
-    } else if (flags['tunnel-url']) {
-      tunnelType = 'custom'
-      tunnel = {
-        mode: 'provided',
-        url: flags['tunnel-url'],
-      }
-    }
+    // getTunnelMode can populate this array. This array is then passed to the dev function.
+    // This allows the dev function to group port warnings into a single renderWarning
+    const portWarnings: PortWarning[] = []
```

Comment:
I don't think you need to start this here and pass it to `getTunnelMode` and the `dev` service.

inside `prepareForDev` before calling `renderPortWarnings` you can just do:
```ts
if (tunnel.mode === 'use-localhost' && tunnel.port !== ports.localhost) {
    portWarnings.push({
      type: 'localhost',
      requestedPort: tunnel.port,
      flag: '--localhost-port',
    })
  }
```